### PR TITLE
fix(common): ensure diffing in ngStyle/ngClass correctly emits value changes

### PR DESCRIPTION
**Pull request by matsko**
Prior to this patch, ngStyle/ngClass would accidentally emit value
changes for static (string-based) values even if the value itself had
not changed. This patch ensures that the style/class diffing code is
more strict and when it signals ngClass/ngStyle that there has been a
value change.

### CODE REVIEW

**Comment by matsko** [comment]
http://test/OCL:286081274:BASE:286081322:1576627681780:768d0322

**Comment by kara** [comment]
@matsko you want this to go into v9, right? I changed the label to "master & patch" but let me know if that's not the case

**Comment by matsko** [comment]
presubmit = https://tap.corp.google.com/ui#id=OCL:286081274:BASE:286310747:1576725141439:7ebd2acb

**Review by mhevery — DISMISSED** [review]
My comments: https://github.com/angular/angular/commit/dcffbdb45abdf4cbe25325f58b786f96bf9f7329

File: packages/common/src/directives/styling_differ.ts [hunk 1]
```diff
@@ -117,25 +157,27 @@ export class StylingDiffer<T> {
...
-          if (this.value) {
-            // we know that the classExp value exists and that it is
-            // a map (otherwise an identity change would have occurred)
-            valueHasChanged = mapHasChanged(keys, this.value as{[key: string]: any}, map);
-          } else {
-            valueHasChanged = true;
-          }
+          // we know that the classExp value exists and that it is
+          // a map (otherwise an identity change would have occurred)
+          valueHasChanged =
```
Comment (antontemchenko) [inline packages/common/src/directives/styling_differ.ts]: inline condition containing booleans always could be simplified. I suggest as below:
```valueHasChanged  = !this.value || mapHasChanged(keys, this.value as{[key: string]: any}, map)```

**Review by IgorMinar — COMMENTED** [review]
I've taken ownership of this PR rebased it and forced-pushed updates

File: packages/common/test/directives/ng_class_spec.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@
  */
 
 import {Component} from '@angular/core';
+import {detectChanges} from '@angular/core/src/render3';
```
Comment (IgorMinar) [inline packages/common/test/directives/ng_class_spec.ts]: oops

**Comment by IgorMinar** [comment]
I ran benchmarks from #34664 on this PR and they show small perf improvement especially for the styling.static_class_and_ngclass_binding.detect_changes and styling.static_class_and_ngstyle_binding_and_style_binding.detect_changes benchmarks

@pkozlowski-opensource I don't believe that the pathological scenario of change-detection that results in no styling/class changes is actually directly tested in those benchmarks. Do you think that it would be adding such scenario just so that we know how much overhead there is in that scenario?

<details>
<summary>The benchmark results are here:</summary>
<p>
test command: `yarn bazel test modules/benchmarks/src/styling/ng2:perf --nocache_test_results` (--config=ivy)
result extraction command: `grep -E "(BENCHMARK|\+-)" dist/out/darwin-fastbuild/testlogs/modules/benchmarks/src/styling/ng2/perf_chromium/test.log`

```
VE (baseline):
          gcAmount |             gcTime |        majorGcTime |     pureScriptTime |         renderTime |         scriptTime
------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------
BENCHMARK styling.no_styling_involved.detect_changes
              0.00 |               0.00 |               0.00 |           1.06+-8% |          5.73+-56% |           1.06+-8%
BENCHMARK styling.static_class.detect_changes
      933.72+-357% |         0.27+-327% |         0.09+-435% |          0.32+-12% |          3.11+-42% |          0.32+-12%
BENCHMARK styling.static_class_with_interpolation.detect_changes
     1070.63+-294% |         0.31+-266% |         0.09+-435% |          1.59+-11% |          3.32+-46% |          1.59+-11%
BENCHMARK styling.class_binding.detect_changes
      831.23+-300% |         0.22+-301% |               0.00 |          1.42+-16% |          2.81+-31% |          1.42+-16%
BENCHMARK styling.static_class_and_class_binding.detect_changes
     1284.06+-277% |         0.36+-250% |         0.08+-435% |          1.77+-15% |          3.30+-43% |          1.77+-15%
BENCHMARK styling.static_class_and_ngclass_binding.detect_changes
     3486.00+-136% |         0.76+-136% |               0.00 |          8.34+-11% |          4.42+-62% |          8.34+-11%
BENCHMARK styling.static_class_and_ngstyle_binding_and_style_binding.detect_changes
     1994.71+-200% |         0.37+-205% |               0.00 |          5.29+-10% |          4.44+-28% |          5.29+-10%


VE (PR@467b3727):
BENCHMARK styling.no_styling_involved.detect_changes
      296.70+-435% |         0.11+-435% |               0.00 |          1.21+-12% |          3.80+-69% |          1.21+-12%
BENCHMARK styling.static_class.detect_changes
      704.95+-435% |         0.16+-435% |         0.07+-435% |          0.32+-18% |          3.16+-48% |          0.32+-18%
BENCHMARK styling.static_class_with_interpolation.detect_changes
     1557.81+-241% |         0.42+-220% |         0.08+-435% |          1.78+-11% |          4.24+-61% |          1.78+-11%
BENCHMARK styling.class_binding.detect_changes
      828.20+-300% |         0.23+-300% |               0.00 |          1.37+-16% |          3.36+-47% |          1.37+-16%
BENCHMARK styling.static_class_and_class_binding.detect_changes
     1211.81+-277% |         0.38+-273% |         0.08+-435% |          1.71+-17% |          4.11+-66% |          1.71+-17%
BENCHMARK styling.static_class_and_ngclass_binding.detect_changes
     2984.98+-152% |         0.63+-155% |               0.00 |           8.12+-6% |          3.44+-44% |           8.12+-6%
BENCHMARK styling.static_class_and_ngstyle_binding_and_style_binding.detect_changes
     1496.23+-239% |         0.26+-246% |               0.00 |          4.64+-14% |          3.50+-42% |          4.64+-14%


Ivy (baseline):
          gcAmount |             gcTime |        majorGcTime |     pureScriptTime |         renderTime |         scriptTime
------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------
BENCHMARK styling.no_styling_involved.detect_changes
     1135.35+-313% |         0.27+-307% |         0.08+-435% |           0.93+-9% |          3.76+-77% |           0.93+-9%
BENCHMARK styling.static_class.detect_changes
              0.00 |               0.00 |               0.00 |           0.26+-8% |          2.85+-39% |           0.26+-8%
BENCHMARK styling.static_class_with_interpolation.detect_changes
      296.91+-300% |         0.19+-300% |               0.00 |           2.36+-7% |          3.36+-53% |           2.36+-7%
BENCHMARK styling.class_binding.detect_changes
      511.03+-329% |         0.17+-300% |               0.00 |          1.47+-15% |          3.32+-49% |          1.47+-15%
BENCHMARK styling.static_class_and_class_binding.detect_changes
      520.33+-328% |         0.17+-300% |               0.00 |          1.74+-13% |          3.77+-36% |          1.74+-13%
BENCHMARK styling.static_class_and_ngclass_binding.detect_changes
     4440.95+-160% |         0.66+-171% |         0.22+-300% |           9.19+-9% |          3.31+-39% |          9.42+-13%
BENCHMARK styling.static_class_and_ngstyle_binding_and_style_binding.detect_changes
     2454.40+-207% |         0.38+-204% |         0.11+-435% |           7.37+-7% |          3.33+-61% |           7.37+-7%


Ivy (PR@467b3727):
          gcAmount |             gcTime |        majorGcTime |     pureScriptTime |         renderTime |         scriptTime
------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------
BENCHMARK styling.no_styling_involved.detect_changes
      558.59+-331% |         0.19+-300% |               0.00 |           1.04+-9% |          5.37+-69% |           1.04+-9%
BENCHMARK styling.static_class.detect_changes
      732.42+-435% |         0.18+-435% |         0.09+-435% |          0.29+-13% |          3.61+-76% |          0.29+-13%
BENCHMARK styling.static_class_with_interpolation.detect_changes
     1085.71+-373% |         0.32+-336% |         0.09+-435% |          2.42+-10% |          3.38+-48% |          2.42+-10%
BENCHMARK styling.class_binding.detect_changes
      803.02+-300% |         0.21+-301% |               0.00 |          1.51+-14% |          4.41+-23% |          1.51+-14%
BENCHMARK styling.static_class_and_class_binding.detect_changes
              0.00 |               0.00 |               0.00 |          1.75+-17% |          3.76+-34% |          1.75+-17%
BENCHMARK styling.static_class_and_ngclass_binding.detect_changes
     2979.25+-182% |         0.52+-177% |         0.23+-302% |          8.34+-11% |          4.01+-34% |          8.34+-11%
BENCHMARK styling.static_class_and_ngstyle_binding_and_style_binding.detect_changes
     1948.15+-200% |         0.52+-237% |               0.00 |          4.93+-12% |          3.73+-43% |          4.93+-12%
```
</p>
</details>

**Comment by IgorMinar** [comment]
blueprint presubmit looks good, but global presubmit seems to be failing with two targets: https://test.corp.google.com/ui#id=OCL:289261315:BASE:289261325:1578771087938:1324b9c3 - we'll need to investigate these failures more.

**Comment by pkozlowski-opensource** [comment]
> I don't believe that the pathological scenario of change-detection that results in no styling/class changes is actually directly tested in those benchmarks. 

@IgorMinar you are right - existing style benchmarks mostly run a scenario where there is an actual change resulting in the DOM write. But the bug being fixed here is of a different kind: no model change with (unnecessary) DOM change.

> Do you think that it would be adding such scenario just so that we know how much overhead there is in that scenario?

Yeh, we could (easily) add such scenario and it would measure "noop" change detection where styling is involved. IMO it would be kind of useful to have it (at the same time we start to have quite a bit of those benchmarks...).

File: packages/common/src/directives/ng_class_impl.ts [hunk 3]
```diff
@@ -152,11 +152,11 @@ export class NgClassR2Impl implements NgClassImpl {
 @Injectable()
 export class NgClassR3Impl implements NgClassImpl {
```
Comment (pkozlowski-opensource) [inline packages/common/src/directives/ng_class_impl.ts]: Shouldn't this be "extends"?

File: packages/common/src/directives/styling_differ.ts [hunk 4]
```diff
@@ -122,96 +124,103 @@ export class StylingDiffer<T> {
...
   }
 
   /**
-   * Examines the last set value to see if there was a change in data.
+   * Examines the last set value to see if there was a change in content.
    *
-   * @param hasIdentityChange whether or not the last set value changed in identity or not
-   *
-   * @returns true when the value has changed (either by identity or by shape if its a collection).
+   * @param inputValueIdentityChanged whether or not the last set value changed in identity or not
```
Comment (matsko) [inline packages/common/src/directives/styling_differ.ts]: add a `.` to the end.

File: packages/common/src/directives/styling_differ.ts [hunk 5]
```diff
@@ -241,48 +250,54 @@ const enum StylingDifferValueTypes {
 
 
 /**
- * builds and returns a map based on the values input value
- *
- * If the `keys` param is provided then the `values` param is treated as a
- * string map. Otherwise `values` is treated as a string array.
+ * @return a normalized string map based on the input string map
```
Comment (matsko) [inline packages/common/src/directives/styling_differ.ts]: we should add comments for each of the `@param` types.

File: packages/common/src/directives/styling_differ.ts [hunk 6]
```diff
@@ -241,48 +250,54 @@ const enum StylingDifferValueTypes {
...
+      // Map uses untrimmed keys, so don't trim until passing to `setMapValues`
+      setMapValues(map, trim ? key.trim() : key, value, parseOutUnits, allowSubKeys);
     }
   }
 
   return map;
 }
 
+/**
+ * @return a normalized string map based on the input string array
```
Comment (matsko) [inline packages/common/src/directives/styling_differ.ts]: add `@param` docs please.

**Review by matsko — COMMENTED** [review]
Looks good. Please look at the small fixes I asked/suggested.

File: packages/common/src/directives/ng_class_impl.ts [hunk 7]
```diff
@@ -168,26 +168,23 @@ export class NgClassR3Impl implements NgClassImpl {
     this._classStringDiffer = this._classStringDiffer ||
         new StylingDiffer('class',
                           StylingDifferOptions.AllowStringValue | StylingDifferOptions.ForceAsMap);
-    this._classStringDiffer.setValue(value);
+    this._classStringDiffer.setInput(value);
   }
 
   setNgClass(value: string|string[]|Set<string>|{[klass: string]: any}) {
```
Comment (pkozlowski-opensource) [inline packages/common/src/directives/ng_class_impl.ts]: I'm surprised that types of this setter are different in `NgClassR3Impl` and `NgClassR2Impl` (IMO `NgClassR2Impl` one is typed incorrectly). I know that this is _not_ about the code you are changing but still trying to understand what is going on here...

File: packages/common/src/directives/ng_class_impl.ts [hunk 8]
```diff
@@ -152,11 +152,11 @@ export class NgClassR2Impl implements NgClassImpl {
 @Injectable()
 export class NgClassR3Impl implements NgClassImpl {
   private _value: {[key: string]: boolean}|null = null;
-  private _ngClassDiffer = new StylingDiffer<{[key: string]: boolean}|null>(
+  private _ngClassDiffer = new StylingDiffer<{[key: string]: true}>(
```
Comment (pkozlowski-opensource) [inline packages/common/src/directives/ng_class_impl.ts]: I'm confused by this change... The way I understand it is that given `<div [ngClass]="exp">` the `exp` can potentially be an object literal with keys being string and values being `any` (if I recall properly we interpret any truthy value as "add class" and any falsy value as "remove class"). As such I would expect T to be `{[klass: string]: any}` (or `{[klass: string]: boolean}` if we normalise things)...

File: packages/common/src/directives/styling_differ.ts [hunk 9]
```diff
@@ -102,145 +124,180 @@ export class StylingDiffer<T> {
...
-        } else {
-          finalValue = tokens.reduce((str, token, i) => str + (i ? ' ' : '') + token);
+      case StylingDifferValueTypes.String: {
+        if (inputValueIdentityChanged) {
+          // process string input only if the identity has changed since the strings are immutable
+          const keys = (this._inputValue as string).split(/\s+/g);
+          if (this._options & StylingDifferOptions.ForceAsMap) {
+            newOutputValue = {} as T;
+            for (let i = 0; i < keys.length; i++) {
+              // TODO remove "as any"
```
Comment (pkozlowski-opensource) [inline packages/common/src/directives/styling_differ.ts]: Are you planning to do this in this PR?

File: packages/common/src/directives/styling_differ.ts [hunk 10]
```diff
@@ -102,145 +124,180 @@ export class StylingDiffer<T> {
...
-    // case 2: array
-    for (let i = 0; i < values.length; i++) {
-      let value = (values as string[])[i];
-      assertValidValue(errorPrefix, value);
-      value = trim ? value.trim() : value;
-      setMapValues(map, value, true, false, allowSubKeys);
+function bulidMapFromStringMap(
+    trim: boolean, parseOutUnits: boolean, allowSubKeys: boolean,
+    values: {[key: string]: string | null | true},
+    keys: string[]): {[key: string]: string | null | true} {
```
Comment (pkozlowski-opensource) [inline packages/common/src/directives/styling_differ.ts]: nit: would it make sense to add explicit type for the return value?

File: packages/common/test/directives/styling_differ_spec.ts [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+      d.setInput(null);
+      expect(d.updateValue()).toBeTruthy();
+      expect(d.value).toEqual(null);
+      expect(d.updateValue()).toBeFalsy();
+      expect(d.value).toEqual(null);
+    });
+  });
+
+
+  describe('hasValueChanged', () => {
```
Comment (pkozlowski-opensource) [inline packages/common/test/directives/styling_differ_spec.ts]: nit: do we want to keep `hasValueChanged` here given that it is gone as a method name?

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM overall, left some comments though and would be to have them acknowledged / addressed (hence the cleanup label).

**Comment by pkozlowski-opensource** [comment]
One more thing: commit(s) message doesn't properly reference issue(s) it is supposed to fix (#34336, #34444)

File: packages/common/src/directives/styling_differ.ts (hunk 9 above)
Comment (IgorMinar) [reply packages/common/src/directives/styling_differ.ts]: not any more, since this code will go away. I'll remove the TODOs

File: packages/common/src/directives/styling_differ.ts (hunk 10 above)
Comment (IgorMinar) [reply packages/common/src/directives/styling_differ.ts]: I'd prefer not to change this any more. If we still needed the differ in the long term, I'd implement the different differently (which would make the types explicit) - one differ interface and two implementations - one differ for styles and one differ for classes. The current code in this PR is hard to follow because it has two many execution branches to accommodate all the scenarios.

File: packages/common/test/directives/styling_differ_spec.ts (hunk 11 above)
Comment (IgorMinar) [reply packages/common/test/directives/styling_differ_spec.ts]: ah.. I should rename it to `updateValue`. will do. thanks for catching this.

File: packages/common/src/directives/ng_class_impl.ts (hunk 3 above)
Comment (IgorMinar) [reply packages/common/src/directives/ng_class_impl.ts]: fixed

File: packages/common/src/directives/styling_differ.ts [hunk 12]
```diff
@@ -24,8 +24,9 @@
  * and unnecessary. Instead, ngClass and ngStyle should have their input values be converted
  * into something that the core-level [style] and [class] bindings understand.
  *
- * This [StylingDiffer] class handles this conversion by creating a new input value each time
- * the inner representation of the binding value have changed.
+ * This [StylingDiffer] class handles this conversion by creating a new output value each time
+ * the input value of the binding value have changed (either via identity change or deep collection
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply packages/common/src/directives/styling_differ.ts]: fixed

File: packages/common/src/directives/styling_differ.ts (hunk 4 above)
Comment (IgorMinar) [reply packages/common/src/directives/styling_differ.ts]: there should be no dots at the end of `@param`, `@return` etc descriptions. I however added one for the line below by accident so I'm removing that instead.

File: packages/common/src/directives/styling_differ.ts (hunk 5 above)
Comment (IgorMinar) [reply packages/common/src/directives/styling_differ.ts]: done

File: packages/common/src/directives/styling_differ.ts (hunk 6 above)
Comment (IgorMinar) [reply packages/common/src/directives/styling_differ.ts]: done

File: packages/common/src/directives/ng_class_impl.ts (hunk 7 above)
Comment (IgorMinar) [reply packages/common/src/directives/ng_class_impl.ts]: you are right. I fixed it. it's odd that typescript didn't complain though.

File: packages/common/src/directives/ng_class_impl.ts (hunk 8 above)
Comment (IgorMinar) [reply packages/common/src/directives/ng_class_impl.ts]: in the implementation itself we only set `true` value when building the map. we never set it to `false` or falsy. So I made the type tighter.

File: packages/common/src/directives/styling_differ.ts (hunk 9 above)
Comment (IgorMinar) [reply packages/common/src/directives/styling_differ.ts]: removed

**Comment by IgorMinar** [comment]
all review feedback processed + PR rebased. I still have one failing target in g3 that I'm investigating

**Comment by IgorMinar** [comment]
@pkozlowski-opensource  I just rebased this on top of the benchmarks that have landed on master

File: integration/_payload-limits.json [hunk 13]
```diff
@@ -39,7 +39,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 2289,
-        "main-es2015": 266648,
+        "main-es2015": 267182,
```
Comment (IgorMinar) [inline integration/_payload-limits.json]: @kara fyi - this regresses on cli-hello-world-lazy by 267182-266869=313 bytes

I don't see an obvious way to cut this back, and since this code is going to be significantly overhauled by misko's PRs let's revisit what the size decrease looks like after his refactor.

**Comment by IgorMinar** [comment]
merge-assistance: global approval
